client/src/app.tsx

```diff
@@ -99,24 +100,36 @@ function selectSpot(spot?: Spot) {
 }
 
 class App extends React.Component<IAppProps, IAppState> {
-    private fabricCount = 0
     private subs: Subscription[] = []
     private orbitDistanceSubject = new BehaviorSubject<OrbitDistance>(OrbitDistance.HELICOPTER)
     private perspectiveCamera: PerspectiveCamera
     private selectedSpotSubject = new BehaviorSubject<Spot | undefined>(undefined)
     private islandState: BehaviorSubject<boolean>
     private physics: Physics
+    private fabricKernel: FabricKernel
+    private instanceUsed: boolean[]
 
     constructor(props: IAppProps) {
         super(props)
         this.physics = new Physics(props.storage)
+        this.physics.applyToFabric(props.fabricExports)
         this.islandState = new BehaviorSubject<boolean>(false)
-        const createGotchiAt = (location: Vector3, jointCountMax: number, genome: Genome): Promise<Gotchi> => {
-            return this.props.createFabricInstance(this.fabricCount++).then(fabricExports => {
-                this.physics.applyToFabric(fabricExports)
-                const fabric = new Fabric(fabricExports, jointCountMax)
-                fabric.createSeed(location.x, location.z)
-                return new Gotchi(fabric, genome)
+        this.fabricKernel = createFabricKernel(props.fabricExports, MAX_POPULATION, INITIAL_JOINT_COUNT)
+        this.instanceUsed = this.fabricKernel.instance.map(() => false)
+        const createGotchiAt = (location: Vector3, genome: Genome): Gotchi => {
+            const freeIndex = this.instanceUsed.indexOf(false)
+            if (freeIndex < 0) {
+                throw new Error("No free fabrics!")
+            }
+            // console.log("Allocating fabric", freeIndex)
+            this.instanceUsed[freeIndex] = true
+            const exports = this.fabricKernel.instance[freeIndex]
+            exports.flushFaces()
+            const fabric = new Fabric(exports, freeIndex).createSeed(location.x, location.z)
+            fabric.iterate(0)
+            return new Gotchi(fabric, genome, (index: number) => {
```

Comment:
What's this constructor arg? You pass in a func which takes and index? Why?

Reply:
The index is its "address" in memory.  0..(N-1)

Reply:
Is there no non-async way to do this?